File: piper/extensions/chrome/src/engine/statements/math_constrain.js
import { stringify } from "../util"


export default async function(statement, onError) {
    try {
        let pointer = statement.pointer
        let value = await this.executeStatement(stringify(pointer.value))
        let minValue = await this.executeStatement(stringify(pointer.min))
        let maxValue = await this.executeStatement(stringify(pointer.max))

        return Math.min(Math.max(value, minValue), maxValue)
    } catch (e) {
        onError(e)
    }
    return null
}

File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/math_constrain.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception
import kotlin.math.max
import kotlin.math.min

suspend fun Piper.executeMathConstrain(statement: JSONObject, onError: (e: Exception) -> Unit): Any? {
    try {
        Log.d("hello", "execute_math_constrain")
        val pointer = statement.getJSONObject("pointer")
        val value = executeStatement(pointer["value"].toString()) ?: return null
        val minValue = executeStatement(pointer["min"].toString()) ?: return null
        val maxValue = executeStatement(pointer["max"].toString()) ?: return null

        val valueDouble = value.toString().toDouble()
        val minDouble = minValue.toString().toDouble()
        val maxDouble = maxValue.toString().toDouble()

        val out = max(min(valueDouble, maxDouble), minDouble)

        return if (out.toString().endsWith(".0")) {
            out.toInt()
        } else {
            out
        }
    } catch (e: Exception) {
        onError(e)
        return null
    }
}